qiskit/optimization/algorithms/recursive_minimum_eigen_optimizer.py

```diff
@@ -14,23 +14,62 @@
 
 """A recursive minimal eigen optimizer in Qiskit's optimization module."""
 
-import logging
 from copy import deepcopy
-from typing import Optional
-
+from enum import Enum
+from typing import Optional, Union, List, Tuple, Dict
+import logging
 import numpy as np
+
 from qiskit.aqua.algorithms import NumPyMinimumEigensolver
 from qiskit.aqua.utils.validation import validate_min
 
-from .minimum_eigen_optimizer import MinimumEigenOptimizer, MinimumEigenOptimizerResult
 from .optimization_algorithm import OptimizationAlgorithm, OptimizationResult
+from .minimum_eigen_optimizer import MinimumEigenOptimizer, MinimumEigenOptimizationResult
 from ..converters.quadratic_program_to_qubo import QuadraticProgramToQubo
 from ..exceptions import QiskitOptimizationError
+from ..problems import Variable
 from ..problems.quadratic_program import QuadraticProgram
 
 logger = logging.getLogger(__name__)
 
 
+class IntermediateResult(Enum):
+    """
+    Defines whether the intermediate results at each iteration should be stored
+    and returned to the end user.
+    """
+
+    NO_ITERATIONS = 0
+    """No intermediate results are stored."""
+
+    LAST_ITERATION = 1
+    """Only results from the last iteration are stored."""
+
+    ALL_ITERATIONS = 2
+    """All intermediate results are stored."""
+
+
+class RecursiveMinimumEigenOptimizationResult(OptimizationResult):
+    """Recursive Eigen Optimizer Result."""
+    def __init__(self, x: Union[List[float], np.ndarray], fval: float,
+                 variables: List[Variable],
+                 replacements: Dict[str, Tuple[str, int]],
+                 history: List[OptimizationResult]) -> None:
+        super().__init__(x, fval, variables, None)
+        self._replacements = replacements
+        self._history = history
+
+    @property
+    def replacements(self) -> Dict[str, Tuple[str, int]]:
+        """Returns replacements."""
+        return self._replacements
+
+    @property
+    def history(self) -> List[OptimizationResult]:
```

Comment:
Since this uses a MinimumEigenOptimizer is there are reason the history is of the base optimization result class type rather than MinimumEigenOptimizationResult?

Reply:
This is a very good question. Yes, initially I started from `List[MinimumEigenOptimizationResult]` but soon I realized I can't do this. The optimizer takes two parameters: `min_eigen_optimizer: MinimumEigenOptimizer` and `min_num_vars_optimizer: Optional[OptimizationAlgorithm]`. The former is used in the loop and the latter when the iteration loop is over. History is collected from the both of them, so the type of a history object is limited by `OptimizationAlgorithm` that's why the list is bounded by `OptimizationResult`.

There are two possible solutions:
1. the PR code.
2. Divide history into two parts: iteration history of type `List[MinimumEigenOptimizationResult]`and the last optimization step of type `OptimizationResult` and expose two properties in the result class of the optimizer.

I was in favor of 1) cause in this case lesser properties and the result looks consistent but lacks of clarity of the `history` property. 2) is better in clarity but if you want to iterate over the history objects, then this requires more code.

So, what do you think?

Reply:
I guess it depends on whether access to variables at the common OptimizationResult is what is being looked for in the history or more specific data. In the latter the specific data depends on actual sub-class type of course. In the case of the OptimizationAlgorithm again I guess there are different return types for each algo. So history is two lists in there the first set of objects being MinEigOptResults and then following them OptResults from the OptAlgo. Are the lists the same length or would one have to go down the list querying each entry for type to see the changeover. Given you know the type of OptAlgo you use once you find them you can access extra fields by the sub-class type you know it will be. Maybe this could be a list or tuple, of 2 lists to at least avoid type checking down a single list. I don't know what code one would want in the normal case to iterate over history - but even if the fields you wanted were that of base class OptResult is there not some semantic differentiation you would want anyway to see when you go from MinEigSolver to OptAlgo in the history since I guess these parts really represent different things/aspects of the overall recursive behavior.

Reply:
Actually history is a list `List[MinimumEigenOptimizationResult]` and a single `OptimizationResult ` we obtain at the last step. Right now all intermediate objects are combined into one list of `OptimizationResult `. I don't know a use case where it is required to iterate over history, but if there's a single list one always knows for sure that the last object in the list may differ from the rest.
In general, I'm fine with a single list or two properties or a tuple of two. So shall we keep as it is now or expand history into two parts?

Reply:
I think my preference would be to have it as a `Tuple[List[MinimumEigenOptimizationResult], OptimizationResult]` which is then clear from the outset. Combining it into a single list seems in my mind to muddy things and I don't see any benefit to a user to do this.

Reply:
Now it is a tuple as a history object.